### DISCUSSION TITLE
Change style of MFXFilterPane

### DISCUSSION THREAD

**Discussion by Roccoxx** [palexdev/MaterialFX#302] [Q&A]
Hi everyone, I'm trying to change the style of a MFXFilterPane in a MFXPaginatedTableView but I don't have luck.

First I have a main class where I set de change of scene with the style :

MFXThemeManager.setOn(pane, Themes.DEFAULT);

```
public void changeStageScene(String path) {
		try {
			FXMLLoader loader = new FXMLLoader();
			InputStream inputStream = getClass().getResourceAsStream(path);
			Parent root = loader.load(inputStream);
			Pane pane = (Pane) root;
			pane.getStylesheets().add(Objects.requireNonNull(getClass().getResource(MainStage.MAIN_STYLE_FILE)).toExternalForm());
			MFXThemeManager.setOn(pane, Themes.DEFAULT);
			MainStage.getInstance().getStage().changeScene(pane);
		}
		catch (Exception e) {
			e.printStackTrace();
		}
	}
```

Later, In my case I use in the view controller for this scene change:

MFXThemeManager.addOn(foodPaginatedTableView, BovinCustomThemes.DEFAULT);

```
public void initialize() {
        foodPaginatedTableView.getSelectionModel().setAllowsMultipleSelection(false);

        MFXThemeManager.addOn(foodPaginatedTableView, BovinCustomThemes.DEFAULT);

        this.createTableColumns();
        this.addTableFilters();
        this.populateTable();

        foodPaginatedTableView.autosizeColumnsOnInitialization();

        When.onChanged(foodPaginatedTableView.currentPageProperty())
                .then((oldValue, newValue) -> foodPaginatedTableView.autosizeColumns())
                .listen();
    }
```
 
This works for me, I can change the style of my MFXTableView fine, but I can't change the style of the MFXFilterPane of it.

I have tried several hours with lookup method in tableview parent, tableview skin, etc and I only found the .default-footer.

Thanks for help me with this.

I use MaterialFX 1.15

**Comment by stefanofornari** [comment]
Hi,
See https://github.com/palexdev/MaterialFX/discussions/294

**Comment by Roccoxx** [comment]
Hi Stefano how are you?

Yes of course, in previus version of MaterialFX (1.13), I made my own table's skin and filterpane's skin, but it's possible in this new themes features of 1.15 version try other way?

**Comment by palexdev** [comment]
If you post a proper issue/answer with all details on what you are trying to achieve, how/what are you trying, then I can try finding a solution that works

**Reply by Roccoxx** [reply]
Sorry, I edited the discussion's content.

**Comment by palexdev** [comment]
@Roccoxx Hey, I figured it out
So, dialogs are kind of a problem when they are used in a Skin because the user has no way to get them unless they create a new Skin to add stylesheets directly onto the dialog (if the dialog is protected)

Also, dialogs are separate scenes and the way we manage it now is to simply add the themes onto them at init time. Which menas they will always be styled but cannot be changed.

I found two ways to fix this:
1) I could add to controls that use dialogs a property that is the factory for the dialog. This way users can change the factory to change the way the dialog is built, which means that you would also be able to add/change/remove their stylesheets
2) "Backport" the style system from the MFXPopups to the dialogs. Popups still work because there is a bridge between them and the control. The bridge automatically adds any stylesheet added onto the control on the popup too.

Do you have any preference on which solution should I implement? I'm kinda undecided


Edit
For instance, to better understand how the two solutions would work:

1)
```java
MFXControl aControl = ...;
aControl.setDialogFactory(() -> ...);
```
2)
```java
MFXControl aControl = ...;
aControl.getStylesheets().add(DIALOG_STYLESHEET);
```

**Comment by Roccoxx** [comment]
Good morning, I not have any preference, but I think it's more easy for use the second way right?

**Reply by palexdev** [reply]
The main difference between the two is that the first solution is slightly more complex/elaborated to use but it's also more flexible as the user can change the way the dialog is built, not only his style, which maybe from another perspective is also breaking encapsulation though
The second solution is easier, much less code but the user can only change the dialog style, nothing else

**Reply by Roccoxx** [reply]
Yes I know, It's probably if I were the creator of a used library like MaterialFX, I would choose the first option. This is because the flexibility could avoid possible issues/discussions in the future.

**Reply by palexdev** [reply]
Yeah I absolutely get the point, I'll think about it
Because the thing that holds me back is:
Take the table as an example, the dialog it is supposed to show is a filter dialog, what if a user builds a generic dialog, that is not the intended behavior, I mean at that point it's the user responsibility, and now that I think about it maybe I could constraint the factory to just generate the dialog's content

About the new version for this enhancement I don't know when it will be out since I'm also waiting for @stefanofornari for making the changes I requested on the recent PR he submitted

**Comment by stefanofornari** [comment]
What about just having the dialog inheriting the stylesheet from its owner instead of using always its own? I saw some controls in a dialog doing it in MaterialFX and I think it is a good compromise that allows users of the table view to fully control the look&feel of all UI belonging to it.

PS: I will provide the changes either today or tomorrow.

**Reply by palexdev** [reply]
> What about just having the dialog inheriting the stylesheet from its owner instead of using always its own? I saw some controls in a dialog doing it in MaterialFX and I think it is a good compromise that allows users of the table view to fully control the look&feel of all UI belonging to it.
> 
> PS: I will provide the changes either today or tomorrow.

Dialogs are containers. What if I have a dialog with other MFX controls in it? Like buttons or check boxes for example. By default dialogs add themes on themselves so that MFX controls are already styled, so the user doesn't have to do anything more than just create and show them.
The filter dialog of the tables as the concrete example. If we don't add the themes on them the filter pane will be un-styled, same if we make it inherit the table' stylesheets, unless the user specifically adds the proper stylesheets on the table
As you may guess at this point, it can get pretty cumbersome for the user

I think the ideal solution is for dialogs to have a default base style and then the possibility of overriding those styles



No worries, I'm not in a hurry

**Reply by palexdev** [reply]
These are the simplest solutions of course.

If I'd have to implement your idea I'd have to change the theme API. Maybe instead of `Theme getTheme()` something like `Set<Theme> getThemes()`

This way the tables would return something like:
```java
public Set<Theme> getThemes() {
    return Set.of(Stylesheets.TABLE_VIEW, Stylesheets.FILTER_DIALOG);
}
```

And even like this I don't think it would completely work, because buttons, fields and other MFX controls in the dialog would still be un-styled, so we still would need to add other Stylesheets just for them.

**Reply by stefanofornari** [reply]
> > What about just having the dialog inheriting the stylesheet from its owner instead of using always its own? I saw some controls in a dialog doing it in MaterialFX and I think it is a good compromise that allows users of the table view to fully control the look&feel of all UI belonging to it.
> > PS: I will provide the changes either today or tomorrow.
> 
> Dialogs are containers. What if I have a dialog with other MFX controls in it? Like buttons or check boxes for example. By default dialogs add themes on themselves so that MFX controls are already styled, so the user doesn't have to do anything more than just create and show them. The filter dialog of the tables as the concrete example. If we don't add the themes on them the filter pane will be un-styled, same if we make it inherit the table' stylesheets, unless the user specifically adds the proper stylesheets on the table As you may guess at this point, it can get pretty cumbersome for the user
> 
> I think the ideal solution is for dialogs to have a default base style and then the possibility of overriding those styles
> 
> No worries, I'm not in a hurry

Of course it should have a default style. What I am suggesting is that the stylesheet should be inherited from the owner if there is one, so the user can customize just one css to change the look&feel of everything. If the owner does not have a stylesheet, the default would apply. IMHO (as a user) it is quite cumbersome as it is now.
